### DISCUSSION TITLE
Deployment of Self-hosting ISR with k18s and docker

### DISCUSSION THREAD

**Discussion by vcernomschi** [vercel/next.js#38858] [Help] [answered]
Hello Community,

I am using Incremental Static Regeneration approach with Next.Js v12.2.

Just wondering if it is possible out of box to specify shared memory location as part of nextJS config.

The context is the following: 
After deployment by using Kubernetes orchestrator with Docker container. In [doc](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr) we have: `
To ensure consistency across all pods, you can disable in-memory caching .... You can use a shared network mount in your Kubernetes pods (or similar setup) to reuse the same file-system cache between different containers. 
`

The key question is:  How can I let the next server know about that shared location, on local it is .next/server/pages

The ideal case it would really amazing to use cloud native storages like AWS S3, Google Bucket and etc.

Thanks in advance,  any links to docs (as I was not able to find this), sample or articles are really welcome

**Comment by kolokytp** [comment]
@vcernomschi have you found a solution to the above?

**Comment by realmunk** [comment]
Any more information regarding this?

**Comment by arealmaas** [comment]
Would be nice with some more documentation about this. We have a setup in Kubernetes with a shared volume now, but are receiving 503's intermittently. Did you land on something @vcernomschi ?

**Reply by garthus** [reply]
Hey @arealmaas 

How did you get this working at all?

I have been able to mount a shared volume, but as its after the original build in the CI/CD deployment - its understandably empty - as the build is done elsewhere.

I have been trying to work out if I need to do another build on each pod when it is created - and the persistent storage is added but cannot work out how.

I would love to know how you managed to get the shared volume working at all if possible?

**Reply by arealmaas** [reply]
Ah I see. The build happens in the Dockerfile and we use an init-container to copy the contents of the Docker-image into the shared volume. Will add some startup time but it's a fair way to get a shared build between pods. Lemme know if you need some examples :)

Where is you build done?

**Reply by garthus** [reply]
Hey @arealmaas 

Oh would love examples - we are still fighting about this.

We are building in BitBucket Pipeline, then pushing to kubernetes via a partner of ours who run this - https://www.section.io/docs/explanations/kubernetes/

We have been fighting it for a while and are still struggling

**Reply by arealmaas** [reply]
Well, start by adding a RWM volume on the deployment:

```yaml
volumes:
        - name: nextjs-shared-storage
          persistentVolumeClaim:
            claimName: efs-pv-claim
```

Add an init-container:

```yaml
initContainers:
        - name: front-end-container
          image: front-end-image
          command:
            [
              "sh",
              "-c",
              "ls /build/BUILD_ID || (cp -r /app/build/* /build/ && echo $POD_TEMPLATE_HASH Succesfully initialised)",
            ]
          volumeMounts:
            - mountPath: /build/
              name: nextjs-shared-storage
              subPathExpr: $(POD_TEMPLATE_HASH)
          env:
            - name: POD_TEMPLATE_HASH
              valueFrom:
                fieldRef:
                  fieldPath: metadata.labels['pod-template-hash']
```

pod-template-hash ensures that new deployments copies new data to the volume so older deployments' data doesn't conflict. 
BUILD_ID ensures that only one pod copies the content. 


Then mount the volume on the next.js-container:

```yaml
containers:
        - name: front-end-container
          image: front-end-image
          volumeMounts:
            - mountPath: /app/.next
              name: nextjs-shared-storage
              subPathExpr: $(POD_TEMPLATE_HASH)
```


In the Dockerfile we have copied the .next lib in the build-directory. Then we are mounting the build-directory as .next in the container.

**Reply by arealmaas** [reply]
Beware, it's a work in progress. As we are still experiencing som weird behaviour of 503's intermittently.

**Reply by garthus** [reply]
hey @arealmaas 

You have been amazing - ok ill see if I can get this up and running and try to help out with what our approach ended up being

**Reply by garthus** [reply]
Hey @arealmaas 

So we have our working now - and have taken a slightly different approach.

Now please forgive me as I am only part of the chain and have much better k8's people in our support.

But what I believe you did is uploading to the next pod.

What we have done is create in each pod a small server which we are calling our uploader - this then is being used by the CI/CD to push the artifacts created in bitbucket

artifacts:
            - .next/**

Into the persistent storage.

That then we are mapping into the mount path

mountPath: /app/.next

which is shared across all the containers in the pod.

This is very much at the limits of my ability / experience however - what we have created now appears to work however, I Am conerned with the race condition for the ISR:

https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr

We have this setup so all pods can write:

  accessModes:
    - ReadWriteMany

In the PVC - so I am not totally sure if this will be ok.

**Reply by arealmaas** [reply]
>however, I Am conerned with the race condition for the ISR

Yes this is our concern as well. Let me know if you get any progress here :)

**Reply by garthus** [reply]
Thanks for all the pointers @arealmaas 

We have been having issues with the init containers wiping out the existing data - so are taking another approach.

We now are looking to push the artifact to s3, then to use an uploader per pod which will upon deployment, pull the data from the s3 bucket.

This we hope will then update each time and should work - this will just be for the built files - the aim is that then the ISR files will be saved as well.

Will see how we go - but wanted to share in case there are red flags

**Comment by aandriamanga** [comment]
Hi, interested with some info about this.

**Comment by garthus** [comment]
If it helps anyone - thanks to @arealmaas - we have a similar approach which we will be taking to production to try out.

What we are doing is this:

- Bitbucket Build
- Bitbucket Push Artifacts to s3 Bucket

Then going through the usual with docker etc. and K8's taking over

In a init container, we are running an s3 sync to pull the files down to the pod - and doing a copy for a few files and the BUILD_ID to 100% make sure all is updated.

This appears to me to be fast, and the s3 sync handles a lot of the issues of files already existing when copied already from other pods.

We are only just getting to production now for a single page on our site (we are a decent size ecommerce site) with the aim now to build from here.

If anyone wants any pointers - please just ask.

We also have varnish in front with a long grace period - which we hope will cover for any small issues with access to pages - but time will tell.

For now, it is building / updateing and ISR appears to be working on a shared drive across the pod

**Comment by mavr1k** [comment]
Working on a similar topic.

I'm using on-demand revalidation and want to have a persistent volume for the revalidated data.
I'm trying to share `.next/static` and `.next/server/pages` but for now I don't know how to copy those folders from the docker image to PersitentVolume.
I think that `initContainers` wouldn't work because I want to run that only once, however, I can make that check as well.

But are there any updates on that? Is there a straightforward way to share and preserve revalidation data across pods?

**Reply by garthus** [reply]
We are using an init container and essentially doing the following:

- Checking if a pod has already synced - by a using the pod_name and creating a file
- Then you have the first container clean up eveything and sync the folders you need
- We then make sure to copy the BUILD_ID in lastn and exit with success

Then in future - any more containers in the same pod will see the pod hash which matches them - and they will check the BUILD_ID matches the BUILD_ID that is there (so if containers spinning up at the same time are blocked from exitign the init

Once the pod is all setup and the build_id is synced, then the init container is done and all the containers can switch to next.js, run it et voila.

Scaling events, the new containers will see they done need to sync again and the ISR is working all this time

**Reply by katsenkatorz** [reply]
hey @garthus thanks for your feedback but

I try this configuration but it does not work, the pods, can you tell me if I am missing something? :

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: site-deployment
  namespace: site-preprod
spec:
  replicas: 3
  template:
    spec:
      initContainers:
        - name: init-cache-sync
          image: alpine:latest
          command: ["/bin/sh", "-c"]
          args:
            - |
              if [ -f "/app/.next/cache-synced" ]; then
                exit 0
              else
                rm -rf /app/.next/*
                cp -r /shared/cache/* /app/.next/
                echo "$BUILD_ID" > /app/.next/lastn
                touch /app/.next/cache-synced
                exit 0
              fi
          volumeMounts:
            - name: cache-volume
              mountPath: /app/.next
      containers:
        - name: site-v3
          image: registry.compagnie.io/compagnie/site-v3:LPOA-2587-fixing-sharing-cache-isr-on-multiple-pod
          volumeMounts:
            - name: cache-volume
              mountPath: /app/.next
      volumes:
        - name: cache-volume
          persistentVolumeClaim:
            claimName: site-v3-next

```

<img width="1435" alt="255865966-bcd21d91-d47f-4d70-9ac8-e80f983e0883" src="https://github.com/vercel/next.js/assets/5629437/5043846d-4101-4cb8-9e85-3d3ef7f5a1d9">

**Reply by garthus** [reply]
Hey @katsenkatorz 

So we had a lot of challenges with this - but here is what we ended up on:

- we are mounting the /app/.next directory as a whole
- we are using bitbucket - and with that, we are building and syncing all files in the ./next folder to an s3 bucket
- Using the s3 script - this is the configmap we have for the init container
- in the s3 bucket we have the environments seperated staging/preprod/production
- I was having issues with the sync and force the CP of the key files at the end
- we are using the Bitbucket Branch as the build_id so its the same across all pods
- We monitor if its not the first container in the pod to see if the build_id is updated and will snooze / exit until the build_id is setup

`
apiVersion: v1
kind: ConfigMap
metadata:
  name: s3-script
data:
  entrypoint.sh: |-
    #!/bin/bash
    POD_NAME=$HOSTNAME
    POD_HASH=${POD_NAME%-${POD_NAME##*-}}
    
    if [ -e "/app/.next/podHash/$POD_HASH" ]; then
      if [ $(ls -1 /app/.next/podHash | wc -l) -gt 1 ]; then
        echo "The /app/.next/podHash folder has multiple commit hash files, and should not happen. Deleting excess files and starting over."
        rm -rf /app/.next/podHash
        exit 1
      fi
      echo "No actions needed. This isn't the first pod of the $POD_HASH replicaset"
    else
      echo "I'm the first pod. I will run s3 sync"
      echo "First, clean up old podHash files"
      rm -rf /app/.next/podHash

      echo "Next, create the files to prevent subsequent pods of the $POD_HASH generation from rerunning sync"
      mkdir /app/.next/podHash
      touch /app/.next/podHash/$POD_HASH

      echo "Setting AWS S3 max concurrent request"
      aws configure set default.s3.max_concurrent_requests 100;
      
      echo "Starting Sync ---------------------"
      aws s3 sync s3://adore-nextjs-sync/{{BITBUCKET_BRANCH}} /app/.next;
      echo "Sync complete ---------------------"

      echo "CP in other specific files along with the BUILD_ID"
      #GW Added the below to be 100% sure these folders get copied in
      echo "CP Static Chunks"
      aws s3 cp s3://adore-nextjs-sync/{{BITBUCKET_BRANCH}} /app/.next/static/chunks --recursive --exclude "*/*"
      echo "CP Static Css"
      aws s3 cp s3://adore-nextjs-sync/{{BITBUCKET_BRANCH}} /app/.next/static/css --recursive --exclude "*/*"
      echo "CP Static Build Folder which is - /app/.next/static/$EXPECTED_BUILD_ID_VALUE"
      aws s3 cp s3://adore-nextjs-sync/{{BITBUCKET_BRANCH}} /app/.next/static/$EXPECTED_BUILD_ID_VALUE --recursive --exclude "*/*"
      echo "Finally the BUILD_ID and the rest"
      aws s3 cp s3://adore-nextjs-sync/{{BITBUCKET_BRANCH}} /app/.next --recursive --exclude "*/*"
      echo "BUILD_ID replaced. Sync complete!! ----------"
    fi

    counter=0
    while [ $counter -lt 30 ] && [ "$(cat /app/.next/BUILD_ID)" != $EXPECTED_BUILD_ID_VALUE ]; do
      echo "BUILD_ID is expected to be $EXPECTED_BUILD_ID_VALUE"
      echo "But BUILD_ID is currently $(cat /app/.next/BUILD_ID). Snoozing 10 seconds."
      counter=$((counter + 1))
      sleep 10
    done

    if [ $counter -ge 30 ]; then
      echo "The loop has reached 30 iterations. Exiting with failure."
      exit 1
    fi

    echo "--------- AWS sync completed, and BUILD_ID is what it should be - $EXPECTED_BUILD_ID_VALUE -----------"`

**Comment by katsenkatorz** [comment]
To solve this, @garthus I wanted to sincerely thank you for your invaluable help in solving the problem. Your suggestion to use initContainers was very helpful and allowed me to find an elegant solution to manage ISR cache sharing between pods, however I didn't have to use Amazon S3, I think it is too much.

After analyzing your S3 configuration, I finally settled on a simpler approach by using initContainers to sync the ISR cache. Here is the final configuration I put in place:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: site-deployment
  namespace: site-preprod
spec:
  replicas: 3
  template:
    spec:
      initContainers:
        - name: init-cache-sync
          image: registry.compagnie.io/myImage
          command: ["/bin/sh", "-c"]
          args:
            - |
              BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
              if [ -d "/app/shared/${BUILD_ID_NEXT}" ]; then
                  echo "Already exist cache ${BUILD_ID_NEXT}"
                  exit 0
              else
                echo "Create cache for ${BUILD_ID_NEXT}"
                rm -rf /app/shared/*
                mkdir -p /app/shared/${BUILD_ID_NEXT}
                cp -r /app/.next/* /app/shared/
                exit 0
              fi
          volumeMounts:
            - name: cache-volume
              mountPath: /app/shared
      containers:
        - name: site-v3
          image: registry.compagnie.io/myImage
          volumeMounts:
            - name: cache-volume
              mountPath: /app/.next
      volumes:
        - name: cache-volume
          persistentVolumeClaim:
            claimName: site-v3-next
```

My approach to cache sharing between pods to optimize cache usage when using Incremental Static Regeneration (ISR) in your Next.js application. Here's how it works:

1. The persistent volume named `cache-volume` which will be used to store the contents of the ".next" folder of my Next.js application.

2. The initContainer is going to be responsible for synchronizing the cache between the pods. The initContainer starts by checking if a specific cache (identified by the build version number - BUILD_ID_NEXT) already exists in the shared folder `/app/ shared`. This is done by looking if there is a folder with the name `BUILD_ID_NEXT` under `/app/shared`. If the cache already exists, that means the sync has already been done for that build version number and the initContainer just exits without doing anything. If the cache does not exist, that means it is the first instance (pod) to start for that specific build version. The initContainer then copies the contents of the app's `.next` folder to `/app/shared/BUILD_ID_NEXT` to store it as a shared cache between Pods.
So when other pods start later for the same build version (same BUILD_ID_NEXT), they will verify that the cache already exists and won't need to sync it again. This saves resources by avoiding duplication of cache generation tasks on each pod.

3.  The main container named `site-v3` to run the Next.js application. This container mounts the "/app/.next" shared volume into the application's `/app/.next` working folder, allowing each pod to access the same shared cache.

In summary, this configuration ensures that the cache generated when using Incremental Static Regeneration (ISR) is shared between pods, which helps to optimize cache usage and reduce redundant build tasks on each pod. This improves the performance and efficiency of your Next.js application when deployed on Kubernetes.

I hope my solution can help someone else here ❤️ !


Regards,

**Reply by garthus** [reply]
Hey @katsenkatorz 

You are welcome and I really want to understand your solution as it looks better than the one I have rolling.

Thanks for sharing this ill have a look when I get the chance

**Reply by garthus** [reply]
hey @katsenkatorz 

Thanks again for this much more elegant solution - I have implemented this my end and it is faster and seems better.

I only have one worry - when see the init container going I seem to see errors like this

rm: can't remove '/app/shared/static/chunks/pages/_app-5494398b97d3a467.js': No such file or directory

Which seems to me that there might be a condition where another init might be running at the same time possibly and the rm is running over each other

**Reply by katsenkatorz** [reply]
Hi @garthus ! 

Thanks for your Feedback,

I didn't get this error, can you show me your config?

**Reply by garthus** [reply]
Hey @katsenkatorz 

thanks for this assistance here is the full config:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: next-deployment
  labels:
    app: next
    environment: production
spec:
  replicas: 5
  selector:
    matchLabels:
      app: next
      environment: production
  template:
    metadata:
      labels:
        app: next
        environment: production
    spec:
      imagePullSecrets:
      - name: dockerhubcredentials
      initContainers:
      - name: s3-sync
        image: {{ACTUAL_IMAGE_NAME}}
        imagePullPolicy: Always
        command: ["/bin/sh", "-c"]
        args:
            - |
              BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
              if [ -d "/app/shared/${BUILD_ID_NEXT}" ]; then
                  echo "Already exist cache ${BUILD_ID_NEXT}"
                  exit 0
              else
                echo "Create cache for ${BUILD_ID_NEXT}"
                rm -rf /app/shared/*
                mkdir -p /app/shared/${BUILD_ID_NEXT}
                cp -r /app/.next/* /app/shared/
                exit 0
              fi
        volumeMounts:
        - name: next-data
          mountPath: /app/shared
        resources:
          requests:
            memory: ".5Gi"
            cpu: "100m"
          limits:
            memory: ".5Gi"
            cpu: "100m"
      containers:
      - name: next
        image: {{ACTUAL_IMAGE_NAME}}
        imagePullPolicy: Always
        resources:
          requests:
            memory: "3000Mi"
            cpu: "1"
          limits:
            memory: "3000Mi"
            cpu: "1"
        ports:
        - containerPort: 3000
        livenessProbe:
          httpGet:
            path: /api/health
            port: 3000
          initialDelaySeconds: 10
          periodSeconds: 5
        readinessProbe:
          httpGet:
            path: /api/health
            port: 3000
          initialDelaySeconds: 5
          periodSeconds: 5
        volumeMounts:
        - name: next-data
          mountPath: /app/.next
      volumes:
      - name: next-data
        persistentVolumeClaim:
          claimName: pvc-next-production
      - name: s3-volume
        configMap:
          defaultMode: 0700
          name: s3-script
```

**Reply by katsenkatorz** [reply]
Hello @garthus , your configuration seems fine to me but I don't know how to help you because I don't reproduce this error :/. I hope you were able to find a solution.

**Reply by garthus** [reply]
Hey @katsenkatorz 

Mate - I think it might be ok - I have not seen similar since.

You have helped me no end and this is the most stable setup we have.

If you are willing - what are your settings / how have you got setup for the following:

- livenessProbe
- readinessProbe

We have setup something but I am not sure if its optimum

**Reply by katsenkatorz** [reply]
Hey @garthus,

From our end, this is how we configured it:

```yaml
//...
volumeMounts:
   - name: cache-volume
      mountPath: /app/.next
livenessProbe:
   httpGet:
      path: /
      port: 3000
      scheme: HTTP
 // ...
 ```
 
 I hope this helps! If you have any other questions or need further details, feel free to ask!

**Reply by MauriceArikoglu** [reply]
Thanks to both of you!!!

**Comment by movergan** [comment]
Thank you for a very details solution @katsenkatorz . I have one question regarding the PVC claim. To be shared among several pods it has to be ReadWriteMany. Am I correct? Could you share a bit more information on that aspect?

**Reply by katsenkatorz** [reply]
With pleasure, @movergan my `PersistentVolumeClaim` (PVC) is currently configured with the access mode `ReadWriteOnce`. Note that not all storage solutions support `ReadWriteMany`. It depends on your underlying storage provider. NFS is one of the storage options that often supports `ReadWriteMany`. 

However I am not a k8s expert, I do not want to mislead you. for the choice of PVC. For me ,`ReadWriteMany`, does not work.

**Reply by garthus** [reply]
Hwy @movergan 

If it helps - I am using ReadWriteMany with NFS and havent seen any issues at all with the ISR at scale so far.

**Reply by garthus** [reply]
Hey @katsenkatorz / @movergan 

If it helps further this at all, I have been working on this and have extended the solution we have and we are now:

On the Init Container 
- Copying across when deploying the new folder which is named with our BUILD_ID
- Deleting the old build folders (we keep the last 4

On the Next Container
- Use a symlink to point to the correct shared folder
- On the old containers - we are linked to the old dir, new containers are the new directory

This means we are never deleting existing data while switching over.

So far this is working out well, and we are trying out but I am seeing this to be the most robust solution.

If anyone wants any direction on how we are doing this, I am happy to share

**Reply by joelim01** [reply]
Hey @garthus. Giant thanks to both you and @katsenkatorz for sharing your work here. We're going through similar steps as we consider moving from SSR to ISG and this has been a huge inspiration for our POC.

I am definitely interested in knowing more about the configuration for the solution you've mentioned.

One other question I have for you is if you've experienced any 404s while transitioning from one deployment to the next using that config and if not, how are you managing your CDN (if you've got one)?

Here is a problem that I'm anticipating:

1. As part of the deploy process, new a new pod is spun up.
2. A user makes a request and hits this new pod before the old pods are terminated. 
3. The server responds with an HTML page (never cached in the CDN) and a list of JS assets for the new deploy. 
4. The browser attempts to load the new JS files. 
5. There is a cache miss in the CDN. 
6. The request is forwarded on to the app server by the CDN.
7. Sometimes, the request for the new JS files will hit one of the old pods. The old pods do not have a copy of that file and the request for that asset will 404 causing the user's browser to fail to load the page.

I've been trying to come up with an adequate solution outside of green blue deploys!

**Reply by garthus** [reply]
Hi @joelim01 

I am really pleased its helping.

So - next.js seems to handle 404 fairly well - as you would expect there is a bit of a switch and to give you context we have layers of varnish cache in front on next.js which will cache for 24 hours the already requested pages.

What I see in real world experience is that when things switch over

- The customers will be getting the old cached content for a while
- There will sometimes be 404 which they see
- In this case, when next gets a 404 for the _ssgManifest.js the js wont load correctly and the page will reload
- This doesnt happen that much as the _ssgManifest.js and the _buildManifest.js are the key items are are heavily cached by varnish
- Also if any customer is hitting the old pods they can get that page and push it to varnish

We have quite a short TTL for our caching the HTML pages so they will go ant get the latest version

When a customer requests a route like:

/_next/data/eb1c809fa0011fb90a59ea53aa8bc587df101bfe/au/c/skin-care.html.json?slug=skin-care.html

We use the build id - so as its switching over, that file might not exist and give a 404

Again, next seems to handle this well as doing a full reload on the route change and as it will hit the new pod it will get the updated data and customers tend to not see any issues.

But yes - you will see the behaviour you expect however, in practice from what I see - you likely want to have your CDN cache for longer than the deployments take to be safe?

And upon switching over, we find it kinda works at least with what we have now.

hope this helps?

**Reply by MauriceArikoglu** [reply]
@garthus can you share the full config you are using now? Including the deletion of old build folders and the symlinking? Thanks!

**Reply by garthus** [reply]
Hey @MauriceArikoglu 

Of course happy to help - so here is the k8's setup for the 2 containers - the s3-sync is the init container which does the copying to the folder which uses the BUILD_ID.

`command: ["/bin/sh", "-c"]
        args:
            - |
              BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
              if [ ! -d /pvc/"${BUILD_ID_NEXT}" ]; then
                  # first init container 
                  echo "OK lets do this - IM FIRST"
                  mkdir -p /pvc/"${BUILD_ID_NEXT}"/ 
                  echo "Now lets Copy to the directory - /pvc/${BUILD_ID_NEXT}/"
                  cp -r /app/.next/* /pvc/"${BUILD_ID_NEXT}"/ 
                  touch /pvc/"${BUILD_ID_NEXT}"/_initcomplete
              fi
              while [ ! -e /pvc/"${BUILD_ID_NEXT}"/_initcomplete ]; do 
              # another init container, wait for copy to complete 
              echo "Im not the first so I will wait till the other is done"
              echo "sleeping for 5"
              sleep 5
              # TODO deal with first copy crashing part-way 
              done 
              echo "Lets Delete the Old Ones"
              find /pvc/ -mindepth 2 -maxdepth 2 -type f -name _initcomplete \
                -exec stat -c %Y$'\t'%n {} \; \
                | sort -nr \
                | tail -n +4 \
                | cut -f2 \
                | cut -d/ -f0-3 \
                | xargs -rtn1 rm -rf`

You can see in the Lets Delete to Old ones this will make sure we dont use too much on the PVC Claim and will keep the 5 most recent builds and delete all older than that - this we feel is a good balance, if we wanted to roll back we can reuse them.

Then the next container is handling the symlink

`command: ["/bin/sh", "-c"]
        args:
            - |
              for target in /app/* /app/.env; do
                lnk=/pvc/"$(basename "${target}")"
                if [ ! -e "${lnk}" ]; then
                  ln -s "${target}" "${lnk}"
                fi
              done
              # point /app/.next at the PVC subfolder for the latest build id
              echo "RM the files"
              rm -rf /app/.next
              echo "Do the Link"
              ln -s /pvc/"$(head -n1 /var/BUILD_ID)"/ /app/.next
              # start node
              exec node server.js`

Hope this helps here - we have one TODO in the init which I am going to deal with today - but this setup is working smoothly for us for now

**edit - sorry for the formatting I Cant seem to get it to format nicely**

**Reply by MauriceArikoglu** [reply]
> Hey @MauriceArikoglu 
> 
> 
> 
> Of course happy to help - so here is the k8's setup for the 2 containers - the s3-sync is the init container which does the copying to the folder which uses the BUILD_ID.
> 
> 
> 
> `command: ["/bin/sh", "-c"]
> 
>         args:
> 
>             - |
> 
>               BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
> 
>               if [ ! -d /pvc/"${BUILD_ID_NEXT}" ]; then
> 
>                   # first init container 
> 
>                   echo "OK lets do this - IM FIRST"
> 
>                   mkdir -p /pvc/"${BUILD_ID_NEXT}"/ 
> 
>                   echo "Now lets Copy to the directory - /pvc/${BUILD_ID_NEXT}/"
> 
>                   cp -r /app/.next/* /pvc/"${BUILD_ID_NEXT}"/ 
> 
>                   touch /pvc/"${BUILD_ID_NEXT}"/_initcomplete
> 
>               fi
> 
>               while [ ! -e /pvc/"${BUILD_ID_NEXT}"/_initcomplete ]; do 
> 
>               # another init container, wait for copy to complete 
> 
>               echo "Im not the first so I will wait till the other is done"
> 
>               echo "sleeping for 5"
> 
>               sleep 5
> 
>               # TODO deal with first copy crashing part-way 
> 
>               done 
> 
>               echo "Lets Delete the Old Ones"
> 
>               find /pvc/ -mindepth 2 -maxdepth 2 -type f -name _initcomplete \
> 
>                 -exec stat -c %Y$'\t'%n {} \; \
> 
>                 | sort -nr \
> 
>                 | tail -n +4 \
> 
>                 | cut -f2 \
> 
>                 | cut -d/ -f0-3 \
> 
>                 | xargs -rtn1 rm -rf`
> 
> 
> 
> You can see in the Lets Delete to Old ones this will make sure we dont use too much on the PVC Claim and will keep the 5 most recent builds and delete all older than that - this we feel is a good balance, if we wanted to roll back we can reuse them.
> 
> 
> 
> Then the next container is handling the symlink
> 
> 
> 
> `command: ["/bin/sh", "-c"]
> 
>         args:
> 
>             - |
> 
>               for target in /app/* /app/.env; do
> 
>                 lnk=/pvc/"$(basename "${target}")"
> 
>                 if [ ! -e "${lnk}" ]; then
> 
>                   ln -s "${target}" "${lnk}"
> 
>                 fi
> 
>               done
> 
>               # point /app/.next at the PVC subfolder for the latest build id
> 
>               echo "RM the files"
> 
>               rm -rf /app/.next
> 
>               echo "Do the Link"
> 
>               ln -s /pvc/"$(head -n1 /var/BUILD_ID)"/ /app/.next
> 
>               # start node
> 
>               exec node server.js`
> 
> 
> 
> Hope this helps here - we have one TODO in the init which I am going to deal with today - but this setup is working smoothly for us for now
> 
> 
> 
> **edit - sorry for the formatting I Cant seem to get it to format nicely**

Hi! Thank you! Mind if I ask if the TODO you mentioned is crucial for the operation and what it is?

I will give this a try, thank you!

```
By the way, you should be able to format multiline code with three backticks the line before and after.
```

**Reply by garthus** [reply]
ooh nice one @MauriceArikoglu 

So i completed the TODO - it wasnt affecting things - but here is the full data now

```
args:
            - |
              echo "Lets Delete the Incomplete Ones"
              find /pvc/ -mindepth 2 -maxdepth 2 -type f -name _initbegin -mmin +60 \
                -exec dirname {} \; \
                | xargs -rt rm -rf
              BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
              if [ ! -d /pvc/"${BUILD_ID_NEXT}" ]; then
                  # first init container 
                  echo "OK lets do this - IM FIRST"
                  mkdir -p /pvc/"${BUILD_ID_NEXT}"/
                  touch /pvc/"${BUILD_ID_NEXT}"/_initbegin
                  echo "Now lets Copy to the directory - /pvc/${BUILD_ID_NEXT}/"
                  cp -r /app/.next/* /pvc/"${BUILD_ID_NEXT}"/ 
                  touch /pvc/"${BUILD_ID_NEXT}"/_initcomplete
                  rm -f /pvc/"${BUILD_ID_NEXT}"/_initbegin
              fi
              begin_ts=$(date +%s)
              while [ ! -e /pvc/"${BUILD_ID_NEXT}"/_initcomplete ]; do
              # another init container, wait for copy to complete 
              echo "Im not the first so I will wait till the other is done"
              echo "sleeping for 5"
              sleep 5
              now_ts=$(date +%s)
              if [ $(( now_ts - begin_ts )) -gt 3600 ]; then
                # waited too long, exit and restart
                exit 1
              fi
              done 
              echo "Lets Delete the Old Ones"
              find /pvc/ -mindepth 2 -maxdepth 2 -type f -name _initcomplete \
                -exec stat -c %Y$'\t'%n {} \; \
                | sort -nr \
                | tail -n +4 \
                | cut -f2 \
                | cut -d/ -f0-3 \
                | xargs -rtn1 rm -rf
```

Let me know how you go and if you need anything else - but this setup we find pretty robust at the moment

**Reply by MauriceArikoglu** [reply]
Lovely @garthus i will get back to you! Thank you so much, let me know if I can reciprocate.

**Comment by MiroStW** [comment]
Hi there, I have another question on this. We have a similar setup in k8s with 2 pods sharing the same cache on a persistedVolume and it works very well. 

There is a problem though (maybe you have stumbled across it too), when a specific path is deleted in the cms and should respond with a 404 from now on. Revalidating the path won't remove it from the cache though, as when response.revalidate('/path/etc') fails, it simply sticks with the last known working version of the path from cache, instead of deleting the cache. That means deleted pages will keep being served for all eternity.

I was hoping there is a better option than to purge the whole cache on demand. Did you come across something like this?

**Reply by MiroStW** [reply]
we now realized we can simply programmatically delete the corresponding files in the cache (at `.next/server/pages/[xyz]`) when a route has been deleted

**Reply by MauriceArikoglu** [reply]
@MiroStW can you share how exactly you delete these pages?

**Reply by MiroStW** [reply]
we are able to simply use an api endpoint within next.js which has access to the file system and delete the files in the aforementioned path. Since the `.next` folder is hosted on a persistedVolume, the file is automatically removed for both pods

**Comment by mackattack3k** [comment]
I also had to set the security context user/group uid as those were set in my Dockerfile as well. This is the complete yaml for me, replaced with dummy values for the name and image:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app
  labels:
    app: my-app
spec:
  replicas: 2
  selector:
    matchLabels:
      app: my-app
  template:
    spec:
      securityContext:
        # Set the same users as in the Dockerfile to avoid permission issues
        runAsUser: 1001
        fsGroup: 1001
      imagePullSecrets:
        - name: gcrsecret
      initContainers:
        - name: init-cache-sync
          image: my-app-image
          # This command will make sure that the build of NextJS is copied into the persistent volume
          command: [ "/bin/sh", "-c" ]
          args:
            - |
              BUILD_ID_NEXT=$(cat /app/.next/BUILD_ID)
              if [ -d "/app/shared/${BUILD_ID_NEXT}" ]; then
                  echo "Already exist cache ${BUILD_ID_NEXT}"
                  exit 0
              else
                echo "Create cache for ${BUILD_ID_NEXT}"
                rm -rf /app/shared/*
                mkdir -p /app/shared/${BUILD_ID_NEXT}
                cp -r /app/.next/* /app/shared/
                echo "Cache created"
                exit 0
              fi
              echo "Delete old cache (keep last 5)"
              ls -t /app/shared/ | tail -n +6 | xargs -I {} rm -rf /app/shared/{}
          volumeMounts:
            - name: my-app--persistent-storage
              mountPath: /app/shared
      containers:
        - name: my-app
          image: my-app-image
          ports:
            - name: http
              containerPort: 3000
          resources:
            limits:
              cpu: 500m
              memory: 512Mi
            requests:
              cpu: 300m
              memory: 258Mi
          envFrom:
            - secretRef:
                name: my-app-env
            - configMapRef:
                name: my-app-env
          livenessProbe:
            httpGet:
              path: /healthz
              port: http
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 5
            timeoutSeconds: 5
          readinessProbe:
            httpGet:
              path: /healthz
              port: http
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 5
            timeoutSeconds: 5
          volumeMounts:
            - name: my-app--persistent-storage
              mountPath: /app/.next
      volumes:
        - name: my-app--persistent-storage
          persistentVolumeClaim:
           claimName: my-app--pvc
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-app--pvc
  labels:
    app: my-app
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
```

**Comment by leerob** [comment, marked as answer]
Hey everyone – we have just published new documentation on self-hosting, including how to configure a custom cache handler for use with ISR / Data Cache in both the Pages and App Router. These docs go in depth about how other features like Image Optimization, Middleware, Environment Variables, Build Caching, and more work when self-hosted. There's also updated Docker based examples.

https://nextjs.org/docs/app/building-your-application/deploying

**Reply by SC0d3r** [reply]
Hi leerob,

I wanted to express my gratitude for providing the documentation. It has been incredibly helpful. I have a question regarding the cache-handler.js file mentioned in the documentation. Specifically, I'm interested in an example of how to write the cache-handler.js file for the /pages directory in my application. Currently, we are calling the API for revalidating ISR, and I was wondering if you could share an example that covers handling revalidateTag in this scenario.

Thank you once again for your support.

**Reply by leerob** [reply]
That code snippet works for the both the Pages and App Routers! E.g. if you do `res.revalidate()` in an API Route, it will also work.

**Reply by pkhadka56** [reply]
Hey,

Do we have an example of disk caching using cache-handler.js? How can we share the cached data among the K8s pods? We generate a few files during `yarn build,` but those files are not used when using Redis. Without Redis, I do not see any issue; however, we need to persist the `.next/cache/` folder. Is it okay if we just persist this folder? We are using SSR, any help will be appreciated. Thank you.
